File: leetcode/math/3009_Maximum_Number_of_Intersections_on_the_Chart/codes/queue.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxIntersectionCount(vector<int>& y) {
        int res = 0;
        vector<pair<long long, long long>> candles;
        for (int i = 0; i + 1 < (int)y.size(); ++i) {
            long long l = 2LL * y[i] + (i ? (y[i] < y[i + 1] ? 1 : -1) : 0);
            long long r = 2LL * y[i + 1];
            candles.push_back({min(l, r), max(l, r)});
        }
        sort(candles.begin(), candles.end());
        priority_queue<long long, vector<long long>, greater<long long>> pq;
        for (auto [l, r] : candles) {
            pq.push(r);
            while (!pq.empty() && pq.top() < l)
                pq.pop();
            res = max(res, (int)pq.size());
        }
        return res;
    }
};

File: leetcode/math/3009_Maximum_Number_of_Intersections_on_the_Chart/codes/queue.py
import heapq

class Solution:
    def maxIntersectionCount(self, y):
        res = 0
        candles = []
        for i in range(len(y) - 1):
            l = 2 * y[i] + (1 if i and y[i] < y[i + 1] else -1 if i else 0)
            r = 2 * y[i + 1]
            candles.append((min(l, r), max(l, r)))
        candles.sort()
        pq = []
        for l, r in candles:
            heapq.heappush(pq, r)
            while pq and pq[0] < l:
                heapq.heappop(pq)
            res = max(res, len(pq))
        return res
